dev.c

```diff
@@ -164,24 +164,31 @@ static void fuse_put_unique(struct fuse_conn *fc, u64 uid)
 	put_cpu();
 }
 
+static struct rdwr_in *fuse_rdwr(struct fuse_conn *fc, uint32_t read)
+{
+	return &fc->queue->requests[read & (FUSE_REQUEST_QUEUE_SIZE - 1)];
+}
+
 static void queue_request(struct fuse_conn *fc, struct fuse_req *req)
 {
-	u32 write, next_index;
-
-	spin_lock(&fc->queue->w.lock);
-	write = fc->queue->w.write;
-	next_index = (write + 1) & (FUSE_REQUEST_QUEUE_SIZE - 1);
-	if (fc->queue->w.read == next_index) {
-		fc->queue->w.read = fc->queue->r.read;
-		BUG_ON(next_index == fc->queue->w.read);
+	u32 write;
+	struct rdwr_in *rdwr;
+	struct fuse_queue_cb *cb = &fc->queue->requests_cb;
+
+	spin_lock(&cb->w.lock);
+	write = cb->w.write;
+	if (write - cb->w.read >= FUSE_REQUEST_QUEUE_SIZE) {
+		cb->w.read = cb->r.read;
+		BUG_ON(write - cb->w.read >= FUSE_REQUEST_QUEUE_SIZE);
```

Comment:
can you please elaborate the fuse_conn (w{read,write} and r{read,write}), as to who(kernel/px-storage) is supposed to update and when, so I understand the direction and definition of this usage correctly.